format/applebookmark/apple_bookmark.go

```diff
@@ -0,0 +1,455 @@
+	d.FieldBool("supports_zero_runs")
+	d.FieldBool("supports_sparse_files")
+	d.FieldBool("is_journaling")
+	d.FieldBool("reserved_3")
+	d.FieldBool("supports_path_from_id")
+	d.FieldBool("supports_mandatory_byte_range_locks")
+	d.FieldBool("supports_hard_links")
+	d.FieldBool("supports_2_tb_file_size")
+
+	d.FieldRawLen("reserved_4", 3)
+	d.FieldBool("has64_bit_object_ids")
+	d.FieldBool("supports_decmp_fs_compression")
+	d.FieldBool("supports_hidden_files")
+	d.FieldBool("supports_remote_events")
+	d.FieldBool("supports_volume_sizes")
+}
+
+func decodePropertyFlags(d *decode.D, bitFn func(d *decode.D)) {
+	d.FieldStruct("property_flags", bitFn)
+
+	d.FieldStruct("enabled_property_flags", bitFn)
+}
+
+var cocoaTimeEpochDate = time.Date(2001, time.January, 1, 0, 0, 0, 0, time.UTC)
+
+type tocHeader struct {
+	tocSize          uint64
+	nextTOCOffset    uint64
+	numEntries       uint64
+	entryArrayOffset int64
+}
+
+func (hdr *tocHeader) decodeEntries(d *decode.D) {
+	for k := uint64(0); k < hdr.numEntries; k++ {
+		entry := new(tocEntry)
+
+		d.FieldStruct("entry", func(d *decode.D) {
+			entry.key = d.FieldU32("key", elementTypeMap)
+
+			// if the key has the top bit set, then (key & 0x7fffffff)
+			// gives the offset of a string record.
+			if entry.key&0x80000000 != 0 {
+				d.FieldStruct("key_string", func(d *decode.D) {
+					d.SeekAbs(calcOffset(entry.key&0x7fffffff), decodeRecord)
+				})
+			}
+
+			entry.recordOffset = calcOffset(d.FieldU32("offset_to_record"))
+
+			d.FieldU32("unused")
+
+			switch entry.key {
+			case elementTypeTargetFlags:
+				d.SeekAbs(entry.recordOffset, resourcePropDecoder.decode)
+			case elementTypeVolumeFlags:
+				d.SeekAbs(entry.recordOffset, volumePropDecoder.decode)
+			default:
+				d.SeekAbs(entry.recordOffset, decodeRecord)
+			}
+
+		})
+	}
+}
+
+func decodeTOCHeader(d *decode.D) *tocHeader {
+	hdr := new(tocHeader)
+
+	d.FieldStruct("toc_header", func(d *decode.D) {
+		hdr.tocSize = d.FieldU32("toc_size")
+		d.FieldU32("magic", d.AssertU(0xfffffffe))
+		d.FieldU32("identifier")
+		hdr.nextTOCOffset = d.FieldU32("next_toc_offset")
+		hdr.numEntries = d.FieldU32("num_entries_in_toc")
+		hdr.entryArrayOffset = d.Pos()
+	})
+
+	return hdr
+}
+
+type tocEntry struct {
+	key          uint64
+	recordOffset int64
+}
+
+const (
+	arrayEntrySize = 4
+	dictEntrySize  = 4
+)
+
+type stack []int64
+
+var offsetStack stack
+
+func (s *stack) pushAndPop(d *decode.D) func() {
+	i := d.Pos()
+	for _, o := range offsetStack {
+		if i == o {
+			d.Fatalf("infinite recursion detected in record decoding")
+		}
+	}
+	*s = append(*s, i)
+	return s.pop
+}
+
+func (s *stack) pop() {
+	*s = (*s)[:len(*s)-1]
+}
+
+func decodeRecord(d *decode.D) {
+	defer offsetStack.pushAndPop(d)()
+
+	d.FieldStruct("record", func(d *decode.D) {
+		n := int(d.FieldU32("length"))
+		typ := d.FieldU32("type", dataTypeMap)
+		switch typ {
+		case dataTypeString:
+			d.FieldUTF8("data", n)
+			d.FieldRawLen("alignment_bytes", 32-((d.Pos()+32)%32))
```

Comment:
Could possibly use `d.AlignBits(32)`

Reply:
I tried that but it was resulting in gaps since fields weren't being produced. Is something like this the way to do it?
```
d.FieldSFn("padding", func(d *decode.D) int64 { return int64(d.AlignBits(32)) })
```

Reply:
Oh sorry now i see, this pads to make the record size an even 4 byte size? `d.FieldRawLen("alignment_bytes", int64(d.AlignBits(32)))` would pad to align to a 32 bit position